feat: Download files from s3.

devarshi-007 opened the pull request:
- Download files from s3 using the button in UI. (S3 File Attachment)
- Download public files in a public folder and private files in the private folder of the site.
- In case of failure of any file while downloading,  It will retry five times for downloading that file.
- Update path -> changes from s3 path to private and public path in database
- It will check whether to download a file or not, using file_url.
- In case of a crash, It will download only files whose URL is AWS URL (private or public) in the database.

## Screenshots

![image](https://github.com/Improwised/frappe_s3_attachment/assets/124968745/37508a1d-f3ee-4339-bd16-6553d846fe53)
![image](https://github.com/Improwised/frappe_s3_attachment/assets/124968745/fce090e6-fcb3-428c-8f87-9c72fcfe8370)

devarshi-007 committed:
feat: Download files from s3

devarshi-007 committed:
feat: Get site path dynamically

devarshi-007 committed:
feat: Update path of database while reverse migrating (#2)

* feat: Update path of database while reverse migrating

* fix: update print to frappe.throw for throwing error

devarshi-007 committed:
fix: Retry in case of file download failure. (#3)

* feat: Update path of database while reverse migrating

* fix: update print to frappe.throw for throwing error

* fix: Retry to download files in case of error

* feat: Checks if file is already exist in local or not

* fix: add frappe._ to frappe.throw in controller.py

Signed-off-by: devarshi-007 <124968745+devarshi-007@users.noreply.github.com>

---------

Signed-off-by: devarshi-007 <124968745+devarshi-007@users.noreply.github.com>

devarshi-007 committed:
fix: Add comment in line 318

Signed-off-by: devarshi-007 <124968745+devarshi-007@users.noreply.github.com>

devarshi-007 committed:
Merge branch 'v13' of https://github.com/Improwised/frappe_s3_attachment into Download_s3

devarshi-007 committed:
fix: Unnecessary space in s3 file attachment doctype

Signed-off-by: devarshi-007 <124968745+devarshi-007@users.noreply.github.com>

devarshi-007 committed:
fix:Give neccessary space in s3 file attachment doctype

Signed-off-by: devarshi-007 <124968745+devarshi-007@users.noreply.github.com>

devarshi-007 committed:
fix: Unnecessary print function

Signed-off-by: devarshi-007 <124968745+devarshi-007@users.noreply.github.com>

munir131 commented:
@devarshi-007 Add instructions in README.md also add command to re install that app as it is already installed on production and that doesn't contain your code.

munir131 commented on frappe_s3_attachment/frappe_s3_attachment/doctype/s3_file_attachment/s3_file_attachment.json:
```diff
@@ -1,447 +1,113 @@
+   "label": "Folder Name"
+  },
   {
-   "allow_bulk_edit": 0, 
-   "allow_in_quick_entry": 0, 
-   "allow_on_submit": 0, 
-   "bold": 0, 
-   "collapsible": 0, 
-   "columns": 0, 
-   "default": "300", 
-   "description": "time should be in seconds", 
-   "fieldname": "signed_url_expiry_time", 
-   "fieldtype": "Int", 
-   "hidden": 0, 
-   "ignore_user_permissions": 0, 
-   "ignore_xss_filter": 0, 
-   "in_filter": 0, 
-   "in_global_search": 0, 
-   "in_list_view": 0, 
-   "in_standard_filter": 0, 
-   "label": "Signed URL expiry time", 
-   "length": 0, 
-   "no_copy": 0, 
-   "permlevel": 0, 
-   "precision": "", 
-   "print_hide": 0, 
-   "print_hide_if_no_value": 0, 
-   "read_only": 0, 
-   "remember_last_selected_value": 0, 
-   "report_hide": 0, 
-   "reqd": 0, 
-   "search_index": 0, 
-   "set_only_once": 0, 
```
@devarshi-007 Are those default values? If you will remove that then will it work as before?

devarshi-007 replied:
It is done by frappe default when I try to add a button. I will check it out.

devarshi-007 committed:
fix: Added dynamic prefix as S3 folder name.

Signed-off-by: devarshi-007 <124968745+devarshi-007@users.noreply.github.com>

devarshi-007 committed:
fix: Update content hash of tabFIle

devarshi-007 committed:
fix: Instructions in README.md

Signed-off-by: devarshi-007 <124968745+devarshi-007@users.noreply.github.com>

devarshi-007 committed:
fix: Markups in README.md

Signed-off-by: devarshi-007 <124968745+devarshi-007@users.noreply.github.com>

devarshi-007 committed:
fix: File path to generate hash

devarshi-007 committed:
fix: change file mode from read to read binary

deltasquare4 commented on frappe_s3_attachment/controller.py:
```diff
@@ -292,6 +305,91 @@ def upload_existing_files_s3(name, file_name):
+# download s3 files
+def download_s3_files(private_local_folder_path, public_local_folder_path):
+    s3_download = S3Operations()
+    bucket = s3_download.S3_RESOURCE.Bucket(s3_download.BUCKET)
```
You are fetching the list of files from wrong place. There should always be a single source of truth, which in this case is the Files doctype. See how the upload function uses it to determine which files to upload. You must do the exact opposite to ensure it works.

In this case, it will break if I have additional files on my S3 bucket that don't exist in Files doctype.

devarshi-007 replied:
It will not break. I have handled that in my code. If any files which entries are not in database will not be downloaded.
For Uploading file they uses file list. But after File uploaded in s3, they modified the file name, attach underscore instead of spaces and attach hash as prefix of file.
For Ex, Tom 22-13.png -> PB7CEXG7_Tom_22-13.png

deltasquare4 commented on frappe_s3_attachment/controller.py:
```diff
@@ -292,6 +305,91 @@ def upload_existing_files_s3(name, file_name):
+    retry_delay = 0.5
+
+    if len(acl.grants) != 1:
+        if acl.grants[1]['Grantee']['Type'] == 'Group' and acl.grants[1]['Grantee']['URI'] == 'http://acs.amazonaws.com/groups/global/AllUsers':
```
Shouldn't this be stored in the Files doctype? Even if it's not, I think we should use the /public path prefix instead of ACL to determine the file's original location.

devarshi-007 replied:
Ok

deltasquare4 commented on frappe_s3_attachment/controller.py:
```diff
@@ -292,6 +305,91 @@ def upload_existing_files_s3(name, file_name):
+# download s3 file
+def download_s3_file(obj, bucket_name, private_local_folder_path, public_local_folder_path):
+    # check from file_url that file is in s3 or local and ommit download if it is in local
+    name = frappe.db.sql(f"""select `name` from `tabFile` where `file_url` LIKE '%{obj.key}%'""")
```
There's a function in original code that matches the file name with URL format. You must use the same function to achieve idempotency instead of using your own.

devarshi-007 replied:
This function will get unique index name from database.
Here, obj.key will also be unique.
The function who matches file url is regex function in controller.py in frappe-s3
It will return only true or false.

devarshi-007 committed:
fix: Use File Doctype for determining file is private or public

devarshi-007 committed:
fix: changes given by Rakshit

devarshi-007 committed:
feat: get object key from file url

devarshi-007 committed:
todo: get details on update_db_s3_to_local

devarshi-007 committed:
fix: changes given by Rakshit

devarshi-007 committed:
Added changes given by Rakshit
1. Files should be match from File list
2. Weather to fetch a file or not is checked using provided regex function
3. This code will update all the table which have url

devarshi-007 committed:
feat:
Added control for Upload function's bug cases
Added functionallity of report generation
